tempodb/encoding/vparquet3/block_autocomplete.go

```diff
@@ -0,0 +1,869 @@
+				continue
+			}
+
+			// Compatible type?
+			typ, _ := c.Type.ToStaticType()
+			if typ == operandType(cond.Operands) {
+				pred, err := createPredicate(cond.Op, cond.Operands)
+				if err != nil {
+					return nil, errors.Wrap(err, "creating predicate")
+				}
+				addPredicate(c.ColumnPath, pred)
+				columnSelectAs[c.ColumnPath] = cond.Attribute.Name
+				continue
+			}
+		}
+
+		// Else: generic attribute lookup
+		genericConditions = append(genericConditions, cond)
+	}
+
+	for columnPath, predicates := range columnPredicates {
+		iters = append(iters, makeIter(columnPath, parquetquery.NewOrPredicate(predicates...), columnSelectAs[columnPath]))
+	}
+
+	attrIter, err := createDistinctAttributeIterator(makeIter, keep, genericConditions, DefinitionLevelResourceAttrs,
+		columnPathResourceAttrKey, columnPathResourceAttrString, columnPathResourceAttrInt, columnPathResourceAttrDouble, columnPathResourceAttrBool, allConditions)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating span attribute iterator")
+	}
+	if attrIter != nil {
+		iters = append(iters, attrIter)
+	}
+
+	batchCol := &distinctBatchCollector{keep: keep}
+
+	var required []parquetquery.Iterator
+
+	// This is an optimization for when all of the resource conditions must be met.
+	// We simply move all iterators into the required list.
+	if allConditions {
+		required = append(required, iters...)
+		iters = nil
+	}
+
+	// Put span iterator last so it is only read when
+	// the resource conditions are met.
+	if spanIterator != nil {
+		required = append(required, spanIterator)
+	}
+
+	// Left join here means the span iterator + 1 are required,
+	// and all other resource conditions are optional. Whatever matches
+	// is returned.
+	return parquetquery.NewLeftJoinIterator(DefinitionLevelResourceSpans, required, iters, batchCol), nil
+}
+
+func createDistinctTraceIterator(makeIter makeIterFn, keep keepFn, resourceIter parquetquery.Iterator, conds []traceql.Condition, start, end uint64, allConditions bool) (parquetquery.Iterator, error) {
+	traceIters := make([]parquetquery.Iterator, 0, 3)
+
+	var err error
+
+	// add conditional iterators first. this way if someone searches for { traceDuration > 1s && span.foo = "bar"} the query will
+	// be sped up by searching for traceDuration first. note that we can only set the predicates if all conditions is true.
+	// otherwise we just pass the info up to the engine to make a choice
+	for _, cond := range conds {
+		switch cond.Attribute.Intrinsic {
+		case traceql.IntrinsicTraceID:
+			traceIters = append(traceIters, makeIter(columnPathTraceID, nil, columnPathTraceID))
+		case traceql.IntrinsicTraceDuration:
+			var pred parquetquery.Predicate
+			if allConditions {
+				pred, err = createIntPredicate(cond.Op, cond.Operands)
+				if err != nil {
+					return nil, err
+				}
+			}
+			traceIters = append(traceIters, makeIter(columnPathDurationNanos, pred, columnPathDurationNanos))
```

Comment:
Using this line to start a discussion but it applies throughout the entire file.  I think there's a further improvement here by using `""` for the last `makeIter` parameter `selectAs`.  SelectAs means the parquet iterators pull back the value, but it's possible to use `selectAs=""` and only assert the predicate.  This will eliminate a lot of allocations and work in the KeepGroup callbacks.  The only one that needs a `selectAs` is the target attribute.   In TraceQL we want to pull back all values, pass them to the engine and the UI but here we don't.

Reply:
Very nice suggestion. Applied! :)